src/nspawn/nspawn.c

```diff
@@ -3058,6 +3130,111 @@ static int setup_uid_map(pid_t pid) {
         return 0;
 }
 
+static int nspawn_dispatch_notify_fd(sd_event_source *source, int fd, uint32_t revents, void *userdata) {
+        _cleanup_fdset_free_ FDSet *fds = NULL;
+        char buf[NOTIFY_BUFFER_MAX+1];
+        struct iovec iovec = {
+                .iov_base = buf,
+                .iov_len = sizeof(buf)-1,
+        };
+        union {
+                struct cmsghdr cmsghdr;
+                uint8_t buf[CMSG_SPACE(sizeof(struct ucred)) +
+                            CMSG_SPACE(sizeof(int) * NOTIFY_FD_MAX)];
+        } control = {};
+        struct msghdr msghdr = {
+                .msg_iov = &iovec,
+                .msg_iovlen = 1,
+                .msg_control = &control,
+                .msg_controllen = sizeof(control),
+        };
+        struct cmsghdr *cmsg;
+        struct ucred *ucred = NULL;
+        int r, *fd_array = NULL;
+        unsigned n_fds = 0;
+        ssize_t n;
+        int inner_child_pid;
+        _cleanup_strv_free_ char **tags = NULL;
+
+        if (!userdata) {
+                return log_error("Not able to retrieve the inner child pid");
+        }
+        inner_child_pid = *(int *)userdata;
```

Comment:
so, instead of passing a ptr to the pid around (and even one with the wrong type, as a pid is of type pid_t, not of int, see below), I think you should cast the pid into a ptr, and pass it around directly. We even have helpers that make that safe for that: PID_TO_PTR() and PTR_TO_PID(), please use that.